DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs

```diff
@@ -358,17 +360,40 @@ CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashSet<strin
                 // This ensures that we've unwound the whole cycle so when we return NoCountInfo.Instance, there will be no cycles in the count info.
                 catch (CyclicalCountElementInfoException ex) when (ex.StartOfCycle == elementName)
                 {
-                    diagnostics.ReportConfigurationNotSupported(marshalUsingData, $"Cyclical {ManualTypeMarshallingHelper.MarshalUsingProperties.CountElementName}");
+                    _diagnostics.ReportConfigurationNotSupported(marshalUsingData, $"Cyclical {ManualTypeMarshallingHelper.MarshalUsingProperties.CountElementName}");
                     return NoCountInfo.Instance;
                 }
             }
 
             return NoCountInfo.Instance;
         }
 
+        private TypePositionInfo? CreateForParamIndex(AttributeData attrData, int paramIndex, ImmutableHashSet<string> inspectedElements)
+        {
+            if (!(_contextSymbol is IMethodSymbol method && 0 <= paramIndex && paramIndex < method.Parameters.Length))
+            {
+                return null;
+            }
+            IParameterSymbol param = method.Parameters[paramIndex];
+            try
+            {
+                return TypePositionInfo.CreateForParameter(
+                    param,
+                    ParseMarshallingInfo(param.Type, param.GetAttributes(), inspectedElements.Add(param.Name)), _compilation) with
+                { ManagedIndex = paramIndex };
+            }
+            // Specifically catch the exception when we're trying to inspect the element that started the cycle.
+            // This ensures that we've unwound the whole cycle so when we return, there will be no cycles in the count info.
+            catch (CyclicalCountElementInfoException ex) when (ex.StartOfCycle == param.Name)
+            {
+                _diagnostics.ReportConfigurationNotSupported(attrData, $"Cyclical {ManualTypeMarshallingHelper.MarshalUsingProperties.CountElementName}");
```

Comment:
Should `Cyclical {0}` be a resource string?

Reply:
Probably. I’ll do a follow-up PR to clean up how diagnostics are reported by the marshalling info parser to be a bit cleaner.